### PR TITLE
[WIP] [CR] Port over ballistic damage type

### PR DESCRIPTION
**Pull request by chaosvolt**
#### Summary

SUMMARY: Features "Port over and implement ballistic damage type"

#### Purpose of change

I've been putting this off for a while and it's gonna be needed for when I finally resume the ammo rebalance project, otherwise rifle turrets are going to be absolutely brutal since rifle AP will be going up.

#### Describe the solution

1. Added `EFFECT_BULLET_IMMUNE` flag. Somehow we'd already picked up the `BULLET_IMMUNE` flag despite it not being used.
2. Converted damage type of ballistic guns and ammo from stab to bullet. Includes paintballs and RPG-7 ammo, oddly they'd converted all the anti-tank weapons but the RPG-7 has stayed stab-type to this very day: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/items/ammo/rpg.json#L17
3. Converted gunmods as needed, laser gunmods correctly converted to modify bullet damage instead of stab damage. Also something not in DDA despite lasers dealing heat damage over there: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/items/gunmod/laser_gunmods.json#L18
4. Added ballistic resist values to materials. These mostly use the values as defined in the original ballistic damage PR, except for some changes aimed towards my stated goal of specific target ballistic defense for certain armor (see below)
6. Added ballistic armor values to monsters. Todo: should add bigger values for monsters that have military armor, which will make arpen from the ammo rebalance more worth it.
7. Updated mods.
8. Updated documentation.
9. Ported over code changes to implement ballistic damage type and resistance.
10. Also ported over fix to bio_ads not actually affecting bullets as intended.

#### Describe alternatives you've considered

Paintball maybe could be bash damage instead? Dunno. Kinda falls into the "why aren't slings and beanbag ammo bash damage" thing too.

#### Testing

1. Checked affected files for syntax and lint errors.
2. Compiled and load-tested.

Testing of materials to hit target armor values put on hold due to an issue I found that I can't figure out the cause of:
image

In a nutshell, ammo that copies from other ammo doesn't seem to actually inherit damage correctly, and I can't figure out what part of the code changes is either broken, not compatable with BN, or not copied right.

I tested both with and without the extra changes to assign.h, no dice.

#### Additional context

Sources:
1. Ballistic damage type, by anothersimulacrum: https://github.com/CleverRaven/Cataclysm-DDA/pull/38912
2. Fix for bio_ads, by swwu: https://github.com/CleverRaven/Cataclysm-DDA/pull/45378

Parts not ported over:
1. Removal of legacy damage support.
2. Some misc. mod updates that involve legacy damage.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
I'll need a good reasoning for it. What does the new type fix?

**Comment by Zireael07** [comment]
> Paintball maybe could be bash damage instead?

Paintball was made bash indeed later since it caused bleeding.

@Coolthulhu: Makes it possible to have modern armor (e.g. ballistic inserts) that protect from bullets, but not so well from stabbing or cutting. Also makes it possible to have monsters that are cut resistant but not bullet resistant or vice versa, so you can have something you can't just mow down in melee or can't just shoot down.

**Comment by chaosvolt** [comment]
> What does the new type fix?

Doing this allows for:
1. Medieval armor to be very effective against cutting monsters and arrow-type projectiles, but vulnerable to bullets.
2. Kevlar to be good against bullets without making them invulnerable to melee weapons.
3. Durable biological monsters can be made hard to kill with melee or arrows while having less armor against bullets.
4. Conversely, kevlar-armored monsters can be made more resistant to bullets without making them immune to melee. So that could be used to make picking good melee weapons against a kevlar hulk risk-reward instead of a terrible idea, along with making kevlar hulks and skeletal juggernauts less functionally identical in terms of gameplay role.

Point two is especially important to the ammo rebalance project, because if I proceed with my planned adjustments to rifle arpen without adjusting ballistic armor first, rifle turrets and rifle-toting NPCs will be absolute hell for the player to deal with.

> Paintball was made bash indeed later since it caused bleeding.

Nice, I'll likely do bash damage conversions in a follow-up PR.

For now, won't be able to take a deeper look at why copy-from is breaking damage values until I get back, will be out today.

**Comment by DrPariah** [comment]
> > What does the new type fix?
> 
> Doing this allows for:
> 
> 1. Medieval armor to be very effective against cutting monsters and arrow-type projectiles, but vulnerable to bullets.
> 2. Kevlar to be good against bullets without making them invulnerable to melee weapons.
> 3. Durable biological monsters can be made hard to kill with melee or arrows while having less armor against bullets.
> 4. Conversely, kevlar-armored monsters can be made more resistant to bullets without making them immune to melee. So that could be used to make picking good melee weapons against a kevlar hulk risk-reward instead of a terrible idea, along with making kevlar hulks and skeletal juggernauts less functionally identical in terms of gameplay role.
> 
> Point two is especially important to the ammo rebalance project, because if I proceed with my planned adjustments to rifle arpen without adjusting ballistic armor first, rifle turrets and rifle-toting NPCs will be absolute hell for the player to deal with.
> 
> > Paintball was made bash indeed later since it caused bleeding.
> 
> Nice, I'll likely do bash damage conversions in a follow-up PR.
> 
> For now, won't be able to take a deeper look at why copy-from is breaking damage values until I get back, will be out today.

So Survivor Plate Armor will suddenly be useful for those of us insane enough to enjoy a good melee build, as opposed to a colorful stopgap measure en route to actual Heavy Survivor Gear?

**Comment by Proxiehunter** [comment]
> > > What does the new type fix?
> > 
> > 
> > Doing this allows for:
> > 
> > 1. Medieval armor to be very effective against cutting monsters and arrow-type projectiles, but vulnerable to bullets.
> > 2. Kevlar to be good against bullets without making them invulnerable to melee weapons.
> > 3. Durable biological monsters can be made hard to kill with melee or arrows while having less armor against bullets.
> > 4. Conversely, kevlar-armored monsters can be made more resistant to bullets without making them immune to melee. So that could be used to make picking good melee weapons against a kevlar hulk risk-reward instead of a terrible idea, along with making kevlar hulks and skeletal juggernauts less functionally identical in terms of gameplay role.
> > 
> > Point two is especially important to the ammo rebalance project, because if I proceed with my planned adjustments to rifle arpen without adjusting ballistic armor first, rifle turrets and rifle-toting NPCs will be absolute hell for the player to deal with.
> > > Paintball was made bash indeed later since it caused bleeding.
> > 
> > 
> > Nice, I'll likely do bash damage conversions in a follow-up PR.
> > For now, won't be able to take a deeper look at why copy-from is breaking damage values until I get back, will be out today.
> 
> So Survivor Plate Armor will suddenly be useful for those of us insane enough to enjoy a good melee build, as opposed to a colorful stopgap measure en route to actual Heavy Survivor Gear?

Is Survivor Plate from a mod? I don't think I've ever seen it and I usually try to rush Survivor gear. Usually it comes down to if I can scavenge enough leather and Kevlar.  I can usually make the duct tape. Once annoyingly I couldn't find a gods damned _soldering iron_ or a book that had the recipe.

**Comment by chaosvolt** [comment]
> Is Survivor Plate from a mod? I don't think I've ever seen it and I usually try to rush Survivor gear.

Yeah, it's from Cata++. Vanilla survivor suits as well will likely lean a bit more towards tolerable ballistic protection too, depending on how things balance out. I've yet to fiddle with material protection just yet due to the above-mentioned bug I ran into.

**Comment by chaosvolt** [comment]
Okay, is it possible that the cause of the problem above is that we might be somehow missing parts of https://github.com/CleverRaven/Cataclysm-DDA/pull/37329 maybe?

**Comment by chaosvolt** [comment]
Closing since source PR author released a port-over that includes an important bugfix I didn't notice, plus successful chery-pick (which I've still been unable to get working).
